src/attr/validators.py

```diff
@@ -63,6 +66,63 @@ def instance_of(type):
     return _InstanceOfValidator(type)
 
 
+@attrs(repr=False, frozen=True)
+class _MatchesReValidator(object):
+    regex = attrib()
+    flags = attrib()
+    func = attrib()
+
+    def __call__(self, inst, attr, value):
+        """
+        We use a callable class to be able to change the ``__repr__``.
+        """
+        if self.func is re.match:
+            match = self.regex.match(value)
+        elif self.func is re.search:
+            match = self.regex.search(value)
+        elif self.func is re.fullmatch:
+            match = self.regex.fullmatch(value)
+        if not match:
+            raise ValueError(
+                "'{name}' must match regex {regex!r}"
+                " ({value!r} doesn't)".format(
+                    name=attr.name, regex=self.regex.pattern, value=value),
+                attr,
+                self.regex,
+                value,
+            )
+
+    def __repr__(self):
+        return "<matches_re validator for pattern {regex!r}>".format(
+            regex=self.regex)
+
+
+# python 2 vs python 3 check if fullmatch is implemented natively
+_native_fullmatch = getattr(re, "fullmatch", None)
+
+
+def matches_re(regex, flags=0, func=_native_fullmatch):
+    """
+    A validator that raises :exc:`ValueError` if the initializer is called
+    with a string that doesn't match the given regex, and :exc:`TypeError`
+    if the initializer is called with a non-string.
+
+    :param regex: a regex string or compiled regex to match against
+    :param flags: flags that will be passed to the underlying re function (default 0)
+    :param func: which underlying re function to call (options are `re.fullmatch()`, `re.search()`, `re.match()`)
```

Comment:
Apart from what flake8 is yelling at you, please also add the default and a `.. versionadded::` tag.

Reply:
👍